File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/argument_builder.rb
require 'cucumber/cucumber_expressions/argument'
require 'cucumber/cucumber_expressions/group'

module Cucumber
  module CucumberExpressions
    class ArgumentBuilder
      def self.build_arguments(regexp, text, parameter_types)
        m = regexp.match(text)
        return nil if m.nil?

        match_group = Group.new(m, text)
        arg_groups = match_group.children

        parameter_types.zip(arg_groups).map do |parameter_type, arg_group|
          groups = nil
          if arg_group
            if arg_group.children.empty?
              groups = [arg_group.value]
            else
              groups = arg_group.children.map {|g| g.value}
            end
          end
          Argument.new(groups, parameter_type)
        end
      end
    end
  end
end

File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ArgumentBuilder.java
package io.cucumber.cucumberexpressions;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class ArgumentBuilder {

    static List<Argument> buildArguments(Pattern pattern, String text, List<ParameterType<?>> parameterTypes) {
        Matcher matcher = pattern.matcher(text);
        if (!matcher.lookingAt()) return null;

        Group matchGroup = new Group(matcher);

        List<Group> argGroups = matchGroup.getChildren();
        if (argGroups.size() != parameterTypes.size()) {
            throw new CucumberExpressionException(String.format("Expression has %s arguments, but there were %s parameterTypes", argGroups.size(), parameterTypes.size()));
        }

        List<Argument> args = new ArrayList<>(argGroups.size());
        for (int i = 0; i < parameterTypes.size(); i++) {
            Group argGroup = i < argGroups.size() ? argGroups.get(i) : null;
            String[] groups;
            if (argGroup != null) {
                if (argGroup.getChildren().isEmpty()) {
                    groups = new String[]{argGroup.getValue()};
                } else {
                    groups = new String[argGroup.getChildren().size()];
                    int g = 0;
                    for (Group group : argGroup.getChildren()) {
                        groups[g++] = group.getValue();
                    }
                }
            } else {
                groups = null;
            }
            ParameterType<?> parameterType = parameterTypes.get(i);
            Argument<?> argument = new Argument<>(groups, parameterType);
            args.add(argument);
        }
        return args;
    }
}
